packages/@ember/-internals/owner/index.ts

```diff
@@ -30,12 +30,15 @@ import EngineInstance from '@ember/engine/instance';
 export interface Owner {
   lookup<T>(fullName: string, options?: LookupOptions): T | undefined;
   factoryFor<T, C>(fullName: string, options?: LookupOptions): Factory<T, C> | undefined;
```

Comment:
Ah, is this where the types are coming from? Ugh. We should think about how to iterate on it, because as it is it lets users of the interface simply lie to the compiler ("this is totally a `Something`, yep, just trust me") and my experience is that those are *often* mistaken: not least because they do not and cannot catch typos. In this case, the types are at least *slightly* safer in that they consistently return `| undefined`, which is the most common kind of error produced by these kinds of types—but this can and will still produce errors down the line. 😩 If we need to ship `@ember/application` supporting this for now and then make a follow-on cross-cutting change to approach it another way, that seems reasonable.

We should also think about the best way to approach this: the use of a `Registry`-style API would solve it, but that has its own tradeoffs and downsides.